### PR TITLE
fast setup for output tensor in tensor iterator

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33165 fast setup for output tensor in tensor iterator**

Adding fast setup support for two more cases besides contiguous:
1. channels last contiguous tensors.
2. non-contiguous but having exactly same shape/stride, no overlapping and dense tensors.
This supported quantized tensor as well.

Benchmark:
```
Before this change:
memory_layout   ms/iter(float)          ms/iter(quant)          quant / float
contiguous      0.02368927001953125     0.027169227600097656    1.1469001610305958
chls_last       0.024966955184936523    0.029038667678833008    1.1630840630640094
chls_last_ish   0.02480936050415039     0.5226361751556396      21.066088143150935
no_fast_setup   0.07995915412902832     0.587709903717041       7.35012657548461

After this change:
memory_layout   ms/iter(float)          ms/iter(quant)          quant / float
contiguous      0.025723934173583984    0.027010679244995117    1.0500213172187518
chls_last       0.02329087257385254     0.027313232421875       1.172701122951407
chls_last_ish   0.023028135299682617    0.027840375900268555    1.2089722219346288
no_fast_setup   0.08249592781066895     0.6507124900817871      7.88781346365599

```
Channels_last_ish memory format got around 20x faster
Benchmark code:

```
import torch
import time
NITER = 1000
def test_fast_setup(memory_format, x, y, qx, qy):
    # Test float
    s = time.time()
    for i in range(NITER):
        x + y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add_relu(qx, qy, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000
    print(memory_format, ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')

print('memory_layout', 'ms/iter(float)', 'ms/iter(quant)', 'quant / float', sep='\t')

x = torch.rand(1, 3, 256, 256)
y = torch.rand(1, 3, 256, 256)
qx = torch.quantize_per_tensor(x, 0.1, 5, torch.quint8)
qy = torch.quantize_per_tensor(y, 0.1, 5, torch.quint8)
test_fast_setup('contiguous', x, y, qx, qy)

cx = x.contiguous(memory_format=torch.channels_last)
cy = y.contiguous(memory_format=torch.channels_last)
cqx = qx.contiguous(memory_format=torch.channels_last)
cqy = qy.contiguous(memory_format=torch.channels_last)
test_fast_setup('chls_last', cx, cy, cqx, cqy)

px = x.permute([0, 2, 3, 1])
py = y.permute([0, 2, 3, 1])
pqx = qx.permute([0, 2, 3, 1])
pqy = qy.permute([0, 2, 3, 1])
test_fast_setup('chls_last_ish', px, py, pqx, pqy)

nx = x.permute([0, 2, 3, 1])
ny = y.permute([0, 3, 2, 1])
nqx = qx.permute([0, 2, 3, 1])
nqy = qy.permute([0, 3, 2, 1])
test_fast_setup('no_fast_setup', nx, ny, nqx, nqy)
```

Differential Revision: [D19825853](https://our.internmc.facebook.com/intern/diff/D19825853)

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.cpp [hunk 1]
```diff
@@ -924,21 +965,67 @@ void TensorIterator::fast_set_up() {
...
+          if (!resize_outputs_ || !op.is_output || op.is_read_write) {
+            return FastSetupType::NONE;
+          }
+          if (!restride_output) {
+            restride_output = true;
+          }
+        }
+      }
+    }
+    if (restride_output) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorIterator.cpp]: compute_fast_setup_type feels like static function, until it is really updates output, this is bad for readability, perhaps you can restride outputs in fast_set_up

Comment (glaringlee) [reply aten/src/ATen/native/TensorIterator.cpp]: Move restride into fast_set_up

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -894,20 +894,67 @@ int TensorIterator::get_dim_to_split() const {
...
+      }
+    case FastSetupType::NON_OVERLAPPING_DENSE:
+      {
+        // find the index of a defined tensor in operands_ start from input tensor
+        int i_defined;
+        for (i_defined = ntensors() - 1; i_defined >= 0; --i_defined) {
+          if (operands_[i_defined].tensor.defined()) break;
+        }
+        TORCH_CHECK(i_defined >= 0, "Can not find a defined tensor when fast allocating memory to outputs");
+        for (int i = 0; i < num_outputs_; i++){
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorIterator.cpp]: ntensors()

File: aten/src/ATen/native/TensorIterator.cpp [hunk 3]
```diff
@@ -924,21 +971,56 @@ void TensorIterator::fast_set_up() {
...
+      is_channels_last &= op.tensor.is_contiguous(at::MemoryFormat::ChannelsLast);
+      is_non_overlapping_and_dense &= op.tensor.is_non_overlapping_and_dense();
+    }
+  }
+
+  if (is_contiguous) {
+    return FastSetupType::CONTIGUOUS;
+  }
+  if (is_channels_last) {
+    if (requires_channels_last_output_) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorIterator.cpp]: This check is unnecessary as `requires_channels_last_output_ ` means any of the inputs are channels_last which is assumed if `is_channels_last == true`

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (glaringlee) [reply aten/src/ATen/native/TensorIterator.cpp]: My bad to give you wrong info before. We should check output only, but in else if statement, I shouldn't check op.is_output again.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**